### ISSUE TITLE
Build language: Rust

### ISSUE THREAD

**Issue by Bert-Proesmans** [HearthSim/Firestarter#2]
This issue is opened to discuss using **Rust** as the programming language for the server implementation. Open a separate issue for promoting another language.

[Rust](https://www.rust-lang.org/) is a relatively new programming language. Rust is statically typed, enforces [move semantics](https://doc.rust-lang.org/rust-by-example/scope/move.html) and [explicit error handling](https://doc.rust-lang.org/book/second-edition/ch09-02-recoverable-errors-with-result.html), this means there are no exceptions like in other languages. Rust has [concurrency constraints](https://doc.rust-lang.org/book/second-edition/ch16-00-concurrency.html) which make it a strict better language than C/C++ for implementing multithreaded programs.
When seriously considering writing software in Rust, it's recommended to start reading [The Book (second edition)](https://doc.rust-lang.org/book/second-edition/index.html).

## Advantages
- Compiler toolchain is easy to install and maintain with [Rustup](https://www.rust-lang.org/en-US/install.html)
- Easy dependency management with [Cargo](https://doc.rust-lang.org/cargo/guide/why-cargo-exists.html) in straightforward syntax (TOML)
- Higher level syntax compiled down into optimized assembly code
- High quality [documentation](https://doc.rust-lang.org/std/) + official [books](https://doc.rust-lang.org/book/) describing complex topics
- Pre-build scripts are written in Rust, no scripting language necessary
- When Rust code compiles, it's guaranteed to behave correctly
- Vibrant/Welcoming community responding quickly to issues; IRC, [Forum](https://users.rust-lang.org/)

## Disadvantages
- Small standard library, using dependencies is encouraged
- Relatively slow compilation times
- Non intuitive syntax (sometimes not ergonomic as well)
- Important libraries are not stable, but usable
- Learning curve
- [The borrow checker](https://doc.rust-lang.org/book/second-edition/ch04-00-understanding-ownership.html)

## Proof of concepts
- https://github.com/Bert-Proesmans/incite
Exploration of possibilities using plain futures. This POC is more complex than it should be!
With:
  - Tokio for asynchronous programming
  - Futures-await to declare futures more ergonomically (requires **Nightly**)
  - Prost for protobuffer encoding/decoding
  - Diesel as type-checked ORM
  - SLog for logging
  - Failure + Error-chain for error handling

### [Tokio](https://github.com/tokio-rs/tokio) VS [Actix](https://github.com/actix/actix)
Actix seems to have a higher throughput for network I/O bound workloads because of thread sticking. Tokio performs true distribution of work across all available threads which results in higher overhead. [source](https://users.rust-lang.org/t/getting-tokio-to-match-actix-web-performance/18659)

**Edit** 2018/07/12: While i do believe that an actor model is a good fit for RPC servers, i'm unsure Actix is currently the way to go. It lacks idiomatic examples of multithreaded programs and is currently facing some hardship because of unhealthy usages of 'unsafe'. 
I'm still doing a deep dive into it's code to understand how it works, maybe we can build a lightweight actor interface for our specific needs. A promising alternative seems to be the [tokio-service](https://github.com/tokio-rs/tokio-service) extension.

**Comment by jleclanche** [comment]
I love Rust as much as the next guy, but is there something wrong with Go? Stove felt pretty good in terms of code health.

**Comment by Bert-Proesmans** [comment]
My biggest issue with Go is the dependency management; imo it's annoying to manually 'go get' them onto your filesystem and then reference them from within each go file.
Cargo does this differently by managing the sources for you after you declared them in one place for your entire program (crate).

Second biggest issue is the implicitness as core principle of the syntax. It makes the code look clean and readable, but in the long term obfuscates what actually happens and it's definitely not beginner friendly. Rust has the latter issue also, but does better on the first.

Go has nice things as well, go routines is one of them but i cannot mention many more things. I didn't really like programming in Go eventually.

**Comment by jleclanche** [comment]
Fair enough. +1 on Rust if we don't go the Go route.

**Comment by Bert-Proesmans** [comment]
Concluding this case with the following;
The required components, as described in the first post, are present and usable.  
**But** Tokio and Futures crates are unstable and change often in preparation of stabilizing features for the first epoch release ([Rust 2018](https://rust-lang-nursery.github.io/edition-guide/2018/status.html)). Since these crates are about asynchronous+I/O code, it affects the aspect of how our code is written.

We should stick to the stable compiler because the surface API of stdlib and dependencies are considered to be stable as well.  
Using the nightly compiler will result in occasions when your code stops compiling the next day. In this case effort must happen to comply with the new nightly release which is not desirable.

Regarding the server code itself, beginning-user-friendliness is an issue. Each new Rust user is pretty soon forced to read the [Rust Book](https://doc.rust-lang.org/book/) to better understand Rust code. Reasons are the young age of the language and it's expressiveness/idiomatic approaches. We can solve this issue with additional documentation, but this is not a sustainable solution.  
Referring to specific sections within the book is a better solution. `You're only a beginner for a short while, but remain an advanced user way longer`.

The above information is based on my experience building out a PoC, which is referred to in the first post.

In short:
- It's doable to create this server and necessary setup code using Rust (only);
- Stick to the stable compiler;
- The epoch 2018 release will come with ergonomics improvements for our case.
